[클린코드 리액트 2기 장태영] 계산기 구현

taeyoungs opened the pull request:
안녕하세요! 🙇‍♂️
클린코드 리액트 2기 수강생 장태영입니다.
잘 부탁드립니다. 🫠

더 늦기 전에 피드백을 한번 받고 싶어서 후다닥 달려왔습니다.

[데모 페이지](https://react-calculator-eosin-nine.vercel.app/)

## 🎯 기능 요구사항

- [x] 2개의 숫자에 대해 덧셈이 가능하다.
- [x] 2개의 숫자에 대해 뺄셈이 가능하다.
- [x] 2개의 숫자에 대해 곱셈이 가능하다.
- [x] 2개의 숫자에 대해 나눗셈이 가능하다.
- [x] AC(All Clear)버튼을 누르면 0으로 초기화 한다.
- [x] 숫자는 한번에 최대 3자리 수까지 입력 가능하다.
- [x] 계산 결과를 표현할 때 소수점 이하는 버림한다.
- [x] 연산의 결과값이 `Infinity`일 경우 `오류`라는 문자열을 보여준다. (아이폰 참고)

## 고려한 점

- 처음에는 이벤트 위임으로 처리를 해보려고 하다가 문득 이런 방식이 리액트를 사용하면서 성능적으로 도움이 되는걸까? 라는 생각이 들었습니다. 평소 리액트를 사용할 때 이벤트 위임으로 처리해본 적이 많이 없기도 해서 이런 부분도 리액트가 도와주고 있을까 싶어서 찾아봤는데 실제로도 그러해서 놀랐습니다. 아직 리액트와 친하지 않은 것 같습니다. 🥲
  - https://dev.to/thawsitt/should-i-use-event-delegation-in-react-nl0
  - https://github.com/facebook/react/issues/13635
- 테스트 코드가 필수 사항은 아니라고 적혀 있긴 했지만 앞으로 진행할 과제들에 비해서 비교적 규모가 작은 계산기에서부터 TDD 사이클을 돌려보고 싶어서 테스트 코드를 작성해봤습니다.
  - 테스트 코드가 뭔가 .. 하드 코딩되어 있다는 느낌을 지울 수가 없네요. 😭
- UI는 UI대로 계산기 관련 로직은 로직대로 분리해서 관리하고 싶어서 계산기 관련 로직을 커스텀 훅으로 분리했습니다.
- reducer를 사용한 이유는 특정 액션으로 상태가 변하게끔 만들고 싶었기 때문입니다. 예를 들어, '연산자를 클릭', '초기화' 등과 같은 액션을 발생시키면 해당 액션에 대한 계산기 로직이 돌아가는 식으로 코드를 작성하고 싶었습니다.
  - 이 부분도 과연 reducer가 필요했을까, 여러 개의 상태값을 지저분하다고 느껴 useReducer를 사용하고 있는건 아닐까와 같은 의문들이 들긴 합니다.

taeyoungs committed:
init: 프로젝트 셋업 with CRA

taeyoungs committed:
chore: 테스팅 툴 관련 ESLint 패키지 및 설정 추가

taeyoungs committed:
test: 사칙연산에 대한 테스트 코드 작성

__test__/calculate.test.tsx
- react-testing-library를 이용하여 덧셈, 뺄셈, 곱셈 그리고 나눗셈
  에 대한 테스트 코드 작성

taeyoungs committed:
chore: 오타 수정

taeyoungs committed:
chore: tsconfig 절대경로 설정

taeyoungs committed:
chore: 파일명 변경

taeyoungs committed:
test: calculate 함수에 대한 테스트 코드 작성

__tests__/calculate.test.ts
- 가산수, 피가산수 그리고 연산자를 파라미터로 받는 calculate
  함수에 대한 테스트 코드 작성
- 덧셈, 뺄셈, 곱셈 그리고 나눗셈에 대한 사칙연산 검증

taeyoungs committed:
feat: calculate 함수 구현

constant.ts
- calculate 함수에서 사용되는 연산자 종류 상수 처리

types.ts
- 특정 객체의 value들을 타입으로 뽑아내는 헬퍼 타입 정의

utils.ts
- 가산수, 피가산수를 연산자 타입에 따라 계산하고 이를 반환하는
  유틸 함수 구현

taeyoungs committed:
mod(calcaulte): 연산자 엣지 케이스 대응

utils.ts
- '+', '-', 'X', '/'를 제외한 연산자를 가지고 calculate 함수를
  호출할 경우에 대한 default 케이스 추가

taeyoungs committed:
chore: 정규식 내 '+' 이스케이프 처리

taeyoungs committed:
feat: Button 컴포넌트 구현

taeyoungs committed:
feat: Calculator 컴포넌트 구현

taeyoungs committed:
test: AC 버튼 동작에 대한 테스트 코드 추가

taeyoungs committed:
mod(Calculator): AC 버튼에 대한 액션 및 핸들러 추가

taeyoungs committed:
test: 자리수에 대한 테스트 코드 작성

taeyoungs committed:
mod(Calculator): 3자리 이상 입력할 경우에 대한 대응

components/Calculator.tsx
- 3자리수 이상 입력하려고 하는 경우 사용자에게 window 기본 알림을
  통하여 불가능함을 알림

taeyoungs committed:
test: 연산 결과값이 Infinity인 경우에 대한 테스트

taeyoungs committed:
mod(Calculator): Inifinity 연산 결과에 대한 대응

components/Calculator.tsx
- 계산기의 연산 결과가 Inifinity일 경우 결과를 보여주는 곳에
  '오류'라는 문자열을 노출

taeyoungs committed:
chore: 에러 메시지 상수 처리

taeyoungs committed:
feat: useCalculator 커스텀 훅 구현

hooks/useCalculator.ts
- 계산기 관련 로직을 커스텀 훅으로 분리

components/Calculator.tsx
- 기존 계산기 관련 로직을 useCalculator 훅에서 반환된 값과 함수로
  대체

taeyoungs committed:
mod: 연산자 입력시 addend 초기화 추가

JunilHwang commented on src/App.tsx:
```diff
@@ -0,0 +1,7 @@
+import Calculator from 'components/Calculator';
```
```suggestion
import { Calculator } from 'components';
```
이런식으로 표현할 수 있도록 index.ts를 활용해주면 좋을 것 같아요!
그러면 import할 때 깔끔하게 관리할 수 있답니다 ㅎㅎ

JunilHwang commented on src/types.ts:
```diff
@@ -0,0 +1,16 @@
+/**
+ * 원하는 객체의 value들을 타입으로 뽑아냅니다.
+ */
+type ValueOf<T extends object> = T[keyof T];
```
주석은 생략해도 무방할 것 같아요!
온전히 이름으로 행위를 표현할 수 있어야 좋은 이름이라고 생각합니다.

JunilHwang commented on src/types.ts:
```diff
@@ -0,0 +1,16 @@
+ */
+type ValueOf<T extends object> = T[keyof T];
+
+export type { Combine, CombineElementProps, ValueOf };
```
```suggestion
/**
 * T와 K에서 T의 프로퍼티를 제거한 타입을 병합합니다.
 */
export type Combine<T, K> = T & Omit<K, keyof T>;
/**
 * 원하는 엘리먼트의 프로퍼티와 커스텀 프로퍼티를 병합합니다.
 */
export type CombineElementProps<E extends ElementType, P = unknown> = Combine<P, ComponentProps<E>>;
/**
 * 원하는 객체의 value들을 타입으로 뽑아냅니다.
 */
export type ValueOf<T extends object> = T[keyof T];
```

이렇게 바로 export를 해주면 네이밍을 변경할 때도 수정을 최소화할 수 있을 것 같아요!

JunilHwang commented on src/utils.ts:
```diff
@@ -0,0 +1,21 @@
+  }
+}
+
+export { calculate };
```
이걸 utils라고 할 수 있을까요?
utils는 "도메인"에 무관하게 쓰일 수 있는 함수들이라고 생각해요.
그런데 이건 완벽하게 도메인에 종속적이네요 ㅎㅎ

taeyoungs replied:
"도메인"에 무관하게 쓰일 수 있는 함수들 .. 앞으로 유틸 함수를 만들어야 한다고 할 때 고려해서 적용하겠습니다!

현재 calculate 함수는 model이라는 키워드를 이용해 CalculatorModel 하위로 옮겼습니다!

JunilHwang commented on src/constant.ts:
```diff
@@ -0,0 +1,29 @@
+  TWO: '2',
+  ONE: '1',
+  ZERO: '0',
+} as const;
```
이럴땐 enum을 써도 되지 않을까요?

taeyoungs replied:
enum으로 변경해봤는데 상당히 좋네요!
enum 사용법에 대해 조금 더 공부해보겠습니다.

JunilHwang commented on src/constant.ts:
```diff
@@ -0,0 +1,29 @@
+  MAX_LENGTH: '숫자는 세 자리까지만 입력 가능합니다!',
+  NUMBER_FIRST: '숫자를 먼저 입력한 후 연산자를 입력해주세요!',
+} as const;
+const MAX_LENGTH = 3;
```
도메인을 다룰 때 필요한 상수와
UI를 다룰 때 필요한 상수가 섞여있는 것 같아요!
이를 분리해서 관리해주면 어떨까요?

taeyoungs replied:
도메인을 다룰 때 필요한 상수라고 생각되는 것들은 contants/calculator.ts로 UI를 다룰 때 필요한 상수라고 생각되는 것들은 message(현재 메시지 관련 밖에 없어서 message라고 해놨습니다)로 분리했습니다!

JunilHwang commented on src/components/Calculator.tsx:
```diff
@@ -0,0 +1,73 @@
+type Operation = ValueOf<typeof OPERATION>;
+
+function Calculator() {
+  const { calculator, calculate, clear, setOperand, setOperation } = useCalculator();
```
지금 components 폴더 밖에 있는 hook을 가져다 쓰고 있어요!
그리고 useCalculator는 현재 컴포넌트(Calulator)에 종속적입니다.
그렇다면, 아예 현재 컴포넌트 하위 폴더에 hooks 폴더를 만들어서 구성해주면 어떨까요?  (응집도를 높이자는 차원입니다!)
```bash
- components
  - calculator
    - hooks
      - useCalculator.ts
    - Calulator.tsx # useCalculator를 불러옵니다.
```

요런 느낌입니다!

JunilHwang commented on src/components/Calculator.tsx:
```diff
@@ -0,0 +1,73 @@
+      return;
+    }
+
+    const digit = e.currentTarget.textContent as Digit;
```
이럴 때는 alias보단 typeGuard 같은걸 사용해야 명확해지는 것 같아요!
```suggestion
    const digit = e.currentTarget.textContent;
    if (isDigit(digit)) {
      setOperand(digit);
    }
```
이러면 runtime에서도 type을 체크할 수 있어서 더 좋다고 생각합니다!

대신 isDigit 라는 함수를 구현해야겠죠?

taeyoungs replied:
typeGuard가 있었네요 .. alias를 쓰면서도 만족스럽지 못했는데 isDigit이라는 typeGuard를 만들어서 활용하니 훨씬 좋은 것 같습니다. :)

JunilHwang commented on src/components/Calculator.tsx:
```diff
@@ -0,0 +1,73 @@
+    if (operation && String(addend).length >= MAX_LENGTH) {
+      alert(ERROR_MESSAGE.MAX_LENGTH);
+      return;
+    }
```
if에 있는 조건을 날것 그대로 쓰기보단, 특정 변수에 담아서 표현해주면 어떨까요?
그리고 또 곰곰이 생각해보면 여기 있는 조건 자체가 도메인로직이기 때문에 아예 함수로 분리해서 관리해주면 더 좋을 것 같아요!

- 연산자가 없을 때는 피연산자를 입력받아야한다.
- 연산자가 있을 때는 연산자를 입력받아야한다.
- 입력받는 수는 3자리 이하여야한다.

위의 3가지가 모두 계산기에 대한 도메인로직 아닐까요!?

JunilHwang replied:
위의 리뷰를 토대로 이야기할 수 있는건,
ui를 다루는 함수에 도메인이 표현되어있고, 이게 뒤섞여있다는 사실입니다!
그래서 요구사항에 변화가 생겼을 때 이 함수가 비대해질 수 있다는거죠.
관리 포인트가 몰리게되고, 그러다보면 점점 코드를 이해하기가 어려워질 수 있지 않을까요?

taeyoungs replied:
전체적으로 도메인 로직과 UI 관련 로직이 뒤섞여 있는 것 같아서 여기에 우선 집중해 코드를 분리했습니다!

JunilHwang commented on src/components/Calculator.tsx:
```diff
@@ -0,0 +1,73 @@
+    }
+
+    setOperation(operation);
+  };
```
이럴 때 이벤트 핸들러의 네이밍은 다음과 같이 표현한답니다!
- handleClickDigit ( digit를 클릭했을 때 처리하는 함수 )
- handleClickOperation ( operation을 click 했을 때 처리하는 함수 )

이렇게 어떤 "행위"에 대한 이벤트 핸들링인지 표현해줘야 좋지 않을까요?
`handleDigit` 처럼 표현한다면, 숫자를 핸들링하겠다는 뜻인데 어떤 행위(이벤트)에 대해 핸들링 하는건지 유추할 수 없어서 불편한 것 같아요!

JunilHwang commented on src/components/Button.tsx:
```diff
@@ -0,0 +1,9 @@
+import type { CombineElementProps } from 'types';
+
+type ButtonProps = CombineElementProps<'button'>;
+
+function Button({ children, ...props }: ButtonProps) {
+  return <button {...props}>{children}</button>;
+}
+
+export default Button;
```
그냥 button을 그대로 쓰는거랑 어떤 차이가 있는걸까요?
컴포넌트로 분리한 이유가 있을까요?

taeyoungs replied:
음 .. '무언가는 컴포넌트로 만들어야 하지 않을까' 하는 생각에 갇혀서 만들어낸 결과물 같습니다.

이 부분은 개인적으로 혼란을 겪고 있는 부분이기도 합니다. 🥲

어느 순간부터 '재사용 가능한 효율적인 컴포넌트'라는 키워드에 생각이 머무르게 됐고
그러다 보니 무언가를 컴포넌트로 만듦에 있어서 
- 이건 과연 재사용이 가능한걸까? 
- 그저 스스로 보기 좋으라고 지저분한 코드를 하나의 컴포넌트로 숨기기 위해서 이 영역을 컴포넌트로 만든건 아닐까? 
- 너무 이르게 컴포넌트로 분리해버린건 아닐까?
- 컴포넌트로 만드는 단위가 이 정도가 맞는건가?

등과 같은 고민을 하게 됐습니다.

문제는 이런 고민들을 해결하지 못하다 보니 이도 저도 아닌 애매한 결과물들을 만들어 내고 있습니다.

예를 들어, 예전의 저라면 다음과 같이 컴포넌트를 분리해놨을 것 같습니다.
계산기에서 결과를 보여주는 영역, 숫자 버튼들이 모여 있는 영역 등 이런 영역을 각각 담당하고 있는 컴포넌트를 만들어야 하고 이를 합쳐서 계산기라는 컴포넌트를 만들어야 한다고 생각했었습니다.

```tsx
<div>
  <Total />
  <Digits />
  <Modifiers />
  <Operations />
</div>
```

그에 반해 지금은 계산기라는 컴포넌트 안에서만 사용되고 단순히 데이터를 받아서 보여주기만 하는 부분을 컴포넌트로 분리하는게 의미가 있는지 .. 다른 영역은 제외하고 Digit 관련 버튼이 모여 있는 부분만 Digits 컴포넌트로 분리했다고 했을 때 Digits라는 컴포넌트 혼자 추상화 단계가 너무 높은건 아닌지 .. 이런 저런 생각에 계속 갇혀 있기만 한 것 같습니다.

말이 너무 길어졌네요!
우선 이 부분은 button을 그대로 사용하는 것이랑 차이가 없는 것 같아서 button을 사용하는 방향으로 가겠습니다. :)

JunilHwang commented on src/components/Calculator.tsx:
```diff
@@ -0,0 +1,73 @@
+      return;
+    }
+
+    setOperation(operation);
```
여기있는 내용들도 앞서 남긴 내용들을 참고해서 개선해주시면 좋을 것 같아요!

JunilHwang commented on src/hooks/useCalculator.ts:
```diff
@@ -0,0 +1,136 @@
+    default:
+      throw new Error('Unhandled Action type');
+  }
+}
```
"계산기"에 대한 도메인 로직이 이렇게 reducer 한 묶음으로 처리되고 있네요!
다만 앞선 리뷰처럼, 여기서 모든걸 처리할 수 없는 상황입니다.
그래서 ui에게 일부 도메인 로직이 끼어들어가있는 상태인거죠.

무엇보다 reducer 라는 함수 하나가 너무 커졌어요.
이렇게 하나의 함수에 포함되는 내용이 많을수록, 이 코드를 보는 사람은 쉽게 이해하기가 힘든 상황입니다 ㅠㅠ

redcuer는 순수하게 값을 받아서 할당하는 역할을 해주고,
값의 연산은 reducer 밖에서 해주면 어떨까요?

즉, 이런 연산들을 reducer가 해주는게 맞을까요?

taeyoungs replied:
제가 작성하면서도 헷갈리는 부분들이 많았는데 확실히 reducer라는 함수가 너무 커지고 도메인 로직과 UI 관련 로직이 섞이면서 보기에도 힘든 상황이 됐던 것 같습니다.

그래서 도메인 로직은 CalculatorModel이라는 파일 하위로 옮기고 이를 통해 reducer는 결과값만을 할당하는 형태로 바꿔보았습니다.

JunilHwang commented on src/__tests__/calculate.test.ts:
```diff
@@ -0,0 +1,30 @@
+
+    expect(calculate(augend, addend, operation)).toBe(1);
+  });
+});
```
지금 제일 중요한 로직은 useCalculator에 있는 reducer가 가지고 있다고 생각하는데, 이에 대한 테스트는 누락되었네요!

JunilHwang commented on src/__tests__/calculator.test.tsx:
```diff
@@ -0,0 +1,125 @@
+
+import App from '../App';
+
+describe('<Calculator />', () => {
```
`Calculator` 컴포넌트가 아닌 `useCalculator` 라는 훅에 대해 테스트하면 더 테스트하기가 쉽지 않을까요?

JunilHwang reviewed (changes requested):
안녕하세요 태영님!
계산기 미션을 함께하게 된 리뷰어 황준일입니다 🙏
일단 코드를 무척 깔끔하게 작성해주신 것 같아요!

지금 더 신경쓰면 좋은 부분은, useCalculator에 있는 reducer의 역할입니다.
reducer 내부에 도메인 로직이 다 들어있는데,
이게 reducer의 역할이 맞을까요?
reducer는 값을 핸들링 해주기만 하면 어떨까요?
무엇보다 태영님이 아닌 다른 사람이 이 코드들을 봤을 때 이해하는데 무척 오래걸릴 것 같아요.

자세한 내용은 상세한 리뷰 내용을 확인해주세요!

그리고 어플리케이션 자체에 대한 내용인데요,

1. 화면에 숫자 0이 있을 때는 연산자 입력이 아예 불가능한 것 같아요.

https://user-images.githubusercontent.com/18749057/218249589-22a9fa86-be9e-4cae-9e2b-35adf10a8cf3.mov

- 연산 결과가 0이 나온 다음에 여기에 다른 연산을 수행할 수도 있지 않을까요?

2. -Infinity

https://user-images.githubusercontent.com/18749057/218249664-51c10ac9-70c2-4ae3-8a0f-c3748c7cdfaa.mov

- Infinity는 오류로 표기되지만, -Infinity는 그대로 출력됩니다.
- "오류"라는 단어가 나온 다음에 숫자 혹은 연산자를 입력하면 그대로 입력되지만, "-Infinity"는 이에 대한 예외 처리가 안 되어있네요!

3. 음수 입력

https://user-images.githubusercontent.com/18749057/218249689-ec161c1f-df09-45e9-99e7-2ae73e93cdd8.mov

이렇게 음수를 입력하고 싶어서 -를 클릭하면 입력이 안 되네요!

이와 비슷하게, `100 / -100` 같은 연산을 하려고 할 때도 문제가 된답니다.

--

테스트코드에 대한 내용인데요,
지금은 `Calculator` 라는 컴포넌트에 대해 테스트를 하고 있어요.
대신에 `useCalculator`에 대해 테스트를 해보면 어떨까요?
그리고 모든 로직을 useCalculator에 위임한다고 했을 때, Calculator에서 전달하는 콜백 함수에 들어있는 도메인 로직도 useCalculator에 위임이 필요하겠죠?
이렇게 해야 UI와 Domain의 완전한 분리가 되는게 아닐까요?

taeyoungs committed:
chore: tsconfig target 옵션 변경(ES6 -> ES2015)

taeyoungs committed:
test: CalculatorModel 클래스에 대한 테스트 추가

taeyoungs committed:
feat: CalculatorModel 클래스 구현

components/Calculator/CalculatorModel.ts
- MVVM 패턴을 참고하여 Model 역할을 담당할 CalculatorModel 클래스
  구현
- 계산기 관련된 데이터를 멤버 변수로 가지고 있으며 멤버 변수를
  업데이트할 수 있는 메서드를 갖고 있는 형태로 구성
- UI 관련 로직은 제외

taeyoungs committed:
feat: 도메인 관련 상수 분리

taeyoungs committed:
test: CalculatorModel 테스트 코드 업데이트

__test__/calculatorModel.test.ts
- 클래스 멤버 변수와 메서드를 테스트하던 테스트 코드에서
  순수 함수를 테스트하는 테스트 코드로 변경

taeyoungs committed:
chore: 불필요해진 테스트 코드 삭제

taeyoungs committed:
mod: CalculatorModel 클래스를 함수 모음으로 변경

CalculatorModel.ts
- 계산기 관련 멤버 변수와 메서드를 가지고 있던 클래스에서
  각 메서드를 일반 함수로 분리한 뒤 이름 모아놓은 파일로
  변경

taeyoungs committed:
chore: 도메인 로직으로 대체된 유틸 함수 삭제

taeyoungs committed:
mod: 상수들을 카테고리에 맞게 분리

taeyoungs committed:
chore: 불필요한 타입 및 컴포넌트 삭제

taeyoungs committed:
feat: Operand 최대 길이 체크 함수 추가

taeyoungs committed:
chore: Operand 최대 길이 상수 처리

taeyoungs committed:
feat: Infinity(or -Infinity) 판별 유틸 함수 추가

taeyoungs committed:
mod: Infinity 상황에 대한 알림 문구 추가

taeyoungs committed:
mod: useCalculator 커스텀 훅 업데이트

useCalculator.ts
- reducer 내 도메인 로직을 CalculatorModel 파일 하위로 분리
- 핸들러 네이밍 변경
- useCalculator로 계산기 관련 로직을 완전히 위임하기 위해
  Calculator 컴포넌트 내에서 정의되어있던 핸들러들을
  useCalculator 훅 내로 이동

Calculator.tsx
- useCalculator 커스텀 훅 업데이트에 따른 수정사항 반영

taeyoungs committed:
chore: Calculator 컴포넌트 관련 파일 이동

taeyoungs committed:
mod: total 변수를 useCalculator 하위로 이동

taeyoungs committed:
chore: named export 추가

taeyoungs committed:
chore: named export 수정

taeyoungs commented:
안녕하세요 준일님!
도메인 로직과 UI 관련 로직을 분리하는 부분에서 생각보다 너무 오래걸렸네요.
언급해주신 방향성을 토대로 나름 수정을 진행했는데 여전히 보기 힘든 코드 같기도 하고 .. 모르겠습니다. 🫠🫠🫠

[데모 페이지](https://react-calculator-eosin-nine.vercel.app/)

## 피드백 사항

### reducer의 역할에 대한 고민

피드백을 받고 제가 작성한 reducer를 들여다 보니 생각보다 더 복잡한 것 같더라구요. 아무래도 도메인 로직과 UI 관련 로직이 뒤섞어 있는 부분과 reducer에서 너무 많은 것을 처리하려고 하다 보니 계속해서 복잡해져만 갔던 것 같습니다.

그래서 도메인 로직을 분리(CalculatorModel)하고 reducer 내에서는 최대한 값을 핸들링만 하는 방향으로 수정을 진행했습니다. 다만, 제 느낌엔 여전히 저만 이해할 수 있는 코드 같기도 합니다. 🥲

++
```typescript
updateAccumulator(updateOperand(state, action.payload));
```
위 코드도 더 좋은 방법이 있지 않았을까 싶지만 작성할 때가 이미 새벽이어서 그런지 더 좋은 방법이 떠오르지 않았네요..

### 그 외

- [x] -Infinity에 대한 처리
- [x] 연산 결과가 0이 나온 다음에 여기에 다른 연산을 수행할 수 있어야 한다.
- [x] 도메인에 종속적인 함수는 utils가 아닌 도메인 관련 함수들이 모여있는 곳에 위치해야 한다.
- [x] 상수를 UI 관련 상수와 도메인 관련 상수로 분리
- [x] 특정 컴포넌트에 종속적인 hooks는 컴포넌트 하위 폴더에 hooks 폴더를 만들어서 구성
- [x] 런타임에서도 타입을 체크할 수 있도록 alias보단 typeGuard 같은걸 사용하자.
- [x] 어떤 "행위"에 대한 이벤트 핸들링인지 표현해주자. (ex. handleClickDigit, handleClickOperation)

## 구현하지 못한 부분

 ### 음수 입력
 
 두 번째 피연산자에 음수를 부여해주는 방법을 고민하고 시도하다가 현재 코드로는 뾰족한 방법이 떠오르지 않아서 구현해보지 못했습니다. '-' 부호를 On/Off 할 수 있게끔 만드려고 했는데 생각처럼 잘 되진 않네요. 😭
 
 ### useCalculator 테스트 코드
 
 Calculator 컴포넌트에 대한 테스트 코드를 useCalculator 훅에 대한 테스트 코드로 변경한다는 걸 깜빡했습니다! 🥲
 이 부분은 따로 진행해보도록 하겠습니다!

JunilHwang commented on src/components/Calculator/index.tsx:
```diff
@@ -0,0 +1,37 @@
+          </button>
+        ))}
+        <button className="operation" onClick={handleClickOperation}>
+          =
```
`=`도 상수로 표현할 수 있을 것 같아요!

JunilHwang commented on src/components/Calculator/CalculatorModel.ts:
```diff
@@ -0,0 +1,84 @@
+  [OPERATION.DIVIDE]: (a: number, b: number) => Math.floor(a / b),
+};
+
+interface ICalculator {
+  operation: Operation | null;
+  leftOperand: number;
+  rightOperand: number | null;
+  accumulator: string;
+}
```
```suggestion
interface ICalculator {
  operation?: Operation;
  leftOperand: number;
  rightOperand?: number;
  accumulator: string;
}
```
이렇게 옵셔널로 표현해줄 수도 있을 것 같아요!
다만 이럴 때는 null이 아니라 undefined로 관리해야겠네요..!

JunilHwang commented on src/components/Calculator/CalculatorModel.ts:
```diff
@@ -0,0 +1,84 @@
+  return isLeftOperandMaxLength || isRightOperandMaxLength;
+};
+
+export { updateAccumulator, updateOperand, updateOperation, calculate, isOperandMaxLength };
+export type { ICalculator };
```
각각의 함수 앞에 export를 붙여주면 자연스럽게 외부에서(다른 모듈에서) 쓰이는 함수인지 아닌지 구분할 수 있어서 좋은 것 같아요!
지금은 최하단에서만 확인할 수 있어서 흐름이 끊기는 느낌입니다.

JunilHwang commented on src/components/Calculator/hooks/index.ts:
```diff
@@ -0,0 +1,3 @@
+import useCalculator from './useCalculator';
+
+export { useCalculator };
```
```suggestion
export { default as userCalculator } from './useCalculator';
```
이렇게 표현할 수 있답니다!

JunilHwang commented on src/__tests__/calculatorModel.test.ts:
```diff
@@ -0,0 +1,97 @@
+  accumulator: '0',
+};
+
+describe('Calculator Model', () => {
+  test('연산자를 저장할 수 있어야 한다.', () => {
+    let calculator = { ...initialState };
```
```suggestion
describe('Calculator Model', () => {
  beforeEach(() => {
      let calculator = { ...initialState };
  });
  
  test('연산자를 저장할 수 있어야 한다.', () => {
```
이렇게 beforeEach에서 각각의 테스트 시작 전에 실행할 코드를 정의할 수 있답니다!

taeyoungs replied:
아 ~ 이런 식으로 쓸 수가 있었군요. 감사합니다! 👍

JunilHwang commented on src/__tests__/calculatorModel.test.ts:
```diff
@@ -0,0 +1,97 @@
+};
+
+describe('Calculator Model', () => {
+  test('연산자를 저장할 수 있어야 한다.', () => {
+    let calculator = { ...initialState };
+
+    calculator = updateOperation(calculator, OPERATION.PLUS);
+
+    expect(calculator.operation).toBe(OPERATION.PLUS);
+  });
```
```suggestion
  test.each([ { operation: OPERATION.PLUS } ])('연산자($operation)를 저장할 수 있어야 한다.', ({ operation }) => {
    let calculator = { ...initialState };

    calculator = updateOperation(calculator, operation);

    expect(calculator.operation).toBe(operation);
  });
```
이런식으로 `each`를 사용하면 테스트에 대한 인자를 여러개 만들어서 사용할 수 있습니다.
이를 이용하여 모든 연산자에 대해 테스트해보면 어떨까요?

taeyoungs replied:
오 .. 정말 좋은 방법 같습니다. 이런 식으로도 테스트를 작성할 수 있네요.
each에 대해 조금 더 알아보고 적용해보겠습니다. 감사합니다. :)

JunilHwang commented on src/__tests__/calculatorModel.test.ts:
```diff
@@ -0,0 +1,97 @@
+    expect(calculator.operation).toBe(OPERATION.PLUS);
+  });
+
+  test('피연산자를 저장할 수 있어야 한다. 단, 오른쪽 피연산자는 연산자의 저장이 선행되어야 한다.', () => {
+    let calculator = { ...initialState };
+
+    calculator = updateOperand(calculator, Digit.SEVEN);
+    expect(calculator.leftOperand).toBe(7);
+
+    calculator = updateOperation(calculator, OPERATION.PLUS);
+
+    calculator = updateOperand(calculator, Digit.FOUR);
+    expect(calculator.rightOperand).toBe(4);
+  });
```
오류를 발생시키는 케이스에 대해서 먼저 테스트를 작성해보면 어떨까요?

JunilHwang commented on src/__tests__/calculatorModel.test.ts:
```diff
@@ -0,0 +1,97 @@
+    expect(calculator.leftOperand).toBe(28);
+  });
+
+  test('나눗셈 연산이 가능해야 한다. 단, 연산 결과의 소수점 이하는 버린다.', () => {
+    let calculator = { ...initialState };
+
+    calculator = updateOperand(calculator, Digit.SEVEN);
+    calculator = updateOperation(calculator, OPERATION.DIVIDE);
+    calculator = updateOperand(calculator, Digit.FOUR);
+    calculator = calculate(calculator);
+
+    expect(calculator.leftOperand).toBe(1);
+  });
```
음수에 대해서도 테스트해보면 어떨까요?
<img width="146" alt="image" src="https://user-images.githubusercontent.com/18749057/218508572-b868c830-673f-46f7-b2f4-a471d7cde8c4.png">

아마 지금 로직으로는 음수에 대해선 버림 연산이 정상적으로 동작하지 않을 것 같아요!

taeyoungs replied:
맞네요!
개인적으로 한번 추가하고 테스트해보겠습니다. ㅎㅎ

JunilHwang reviewed (approved):
안녕하세요 태영님!
리팩토링 잘 진행해주셨네요 👍👏👏
특히 테스트 코드를 꼼꼼하게 작성해주셔서 좋았어요!
다만 더 다양한 케이스에 대해 테스트 해보면 좋을 것 같습니다 ㅎㅎ
이럴 때 `test.each`를 사용하면 좋습니다.

이 외에도 자잘한 리뷰 남겼으니 한 번 확인 부탁드려요!
별개로 계산기 미션은 온보딩 미션이기 때문에 여기서 마무리하도록 하겠습니다 😁

저의 피드백들이 태영님께 조금이나마 도움이 되었길 바랍니다!
나머지 미션도 무사완주 하시길 🙏

감사합니다! 화이팅 💪💪
